장바구니기능 구현

tlqkrus012345 opened the pull request.

tlqkrus012345 committed:
feat: 장바구니 컨트롤러

- crud 구현

Resolves: #8

tlqkrus012345 committed:
feat: 장바구니 서비스

- redis hash 자료구조를 사용하여 장바구니 구현

Resolves: #8

tlqkrus012345 committed:
test: 장바구니 서비스

- redis에 실제로 저장되는지 테스트를 위해 통합 테스트로 진행

Resolves: #8

tlqkrus012345 committed:
refacotr: security url

- cart url 허용

Resolves: #8

tlqkrus012345 committed:
feat: cart http

- crud

Resolves: #8

tlqkrus012345 committed:
refactor: 상수 위치 변경

- SessionLoginService에 책임 부여

Resolves: #8

tlqkrus012345 committed:
refactor: 매개변수를 통해 세션에 할당

- 로그인 정보, 장바구니 사용하기 위해 매겨변수를 통해 사용할 수 있게 변경

Resolves: #8

tlqkrus012345 committed:
refactor: security 세션 생성

- never 로 변경

Resolves: #8

tlqkrus012345 committed:
feat: 레디스 설정

- 로그인 정보 직렬화
- Lettuce 명시화

Resolves: #8

tlqkrus012345 committed:
refacotr: HttpSession에 저장하도록 변경

Resolves: #8

tlqkrus012345 committed:
refactor: Map -> List

- 응답 구조 명확하게 확인할 수 있도록 변경

Resolves: #8

tlqkrus012345 committed:
test: 테스트 수정

- 테스트 용 로그인 상수 적용
- 장바구니에 추가 메서드 이름 변경

Resolves: #8

tlqkrus012345 committed:
rename: 파일명

Resolves: #8

tlqkrus012345 committed:
feat: cart.http

- 수량 조정 추가

Resolves: #8

f-lab-martin commented on src/main/java/com/paytogether/cart/controller/CartController.java:
```diff
@@ -0,0 +1,58 @@
+
+  private final CartService cartService;
+
+  @PostMapping("/cart/add/{memberId}")
```
url path를 어떻게 맞추면 path의 파라미터 의미들을 알기 좋을까요?
Restful한 API 설계에 대해서 공부해보시고 Restful하게 수정해보시길 바랍니다 :)

tlqkrus012345 replied:
path 파라미터 의미들을 알기 좋을 수 있게 회원의 장바구니 의미를 표현할 수 있게 "/members/{memberId}/carts" 로 변경했습니다.

f-lab-martin commented on src/main/java/com/paytogether/cart/controller/CartController.java:
```diff
@@ -0,0 +1,58 @@
+    return ApiResponse.createResponse(HttpStatus.OK.value(), CartResponse.of(cartResult.getCart()));
+  }
+
+  @DeleteMapping("/cart/delete/{memberId}/{productId}")
```
Http Method로 DELETE를 사용했으니 delete가 path에서 빠지면 어떨까요?

tlqkrus012345 replied:
Restful한 API 설계 원칙을 준수하기 위해 delete 동사를 제거했습니다!

f-lab-martin commented on src/main/java/com/paytogether/cart/controller/CartController.java:
```diff
@@ -0,0 +1,58 @@
+    return ApiResponse.createResponse(HttpStatus.OK.value(), CartResponse.of(cartResult.getCart()));
+  }
+
+  @DeleteMapping("/cart/clear/{memberId}")
```
clear는 Delete보다는 PUT에 가까울 것 같습니다. 다시 고민해보시면 어떨까요?

tlqkrus012345 replied:
장바구니 전체 데이터를 업데이트하는 의미를 명확히 하기 위해 해당 API의 HTTP 메서드를 PUT으로 변경했습니다!

f-lab-martin commented on src/main/java/com/paytogether/cart/controller/request/ProductRequest.java:
```diff
@@ -0,0 +1,9 @@
+import lombok.Getter;
+
+@Getter
+public class ProductRequest {
```
java record를 사용해서 DTO를 표현해보면 어떨까요?

그리고 ProductRequest는 정확히 어떤 걸 요청하는 클래스일까요?

tlqkrus012345 replied:
ProductRequest는 상품 번호와 수량을 클라이언트로부터 받기 위한 DTO 클래스였습니다. 해당 클래스의 용도를 명확히 하기 위해 AddProductRequest로 이름을 변경했습니다!

f-lab-martin replied:
기존에 작성했던 다른 DTO들도 record로 바꿀 수 있는지 다시 검토해보면 좋을 것 같습니다~!

f-lab-martin commented on src/main/java/com/paytogether/cart/controller/response/CartResponse.java:
```diff
@@ -0,0 +1,19 @@
+@Getter
+public class CartResponse {
+
+  private List<CartDto> cart;
```
cart list를 표현하는 변수명도 동일하게 cart로 표현되는걸까요?

tlqkrus012345 replied:
처음에는 장바구니 전체를 하나로 표현하려고 Cart라는 이름을 사용했지만 실제로는 CartDto 객체들의 리스트 형태이므로, 의미를 더 명확히 하기 위해 cartDtoList로 변수명을 변경했습니다!

f-lab-martin replied:
cart는 장바구니인데, 결국 장바구니에 <상품, 상품개수>가 페어로 쌓이게 됩니다. 그러면 List<Pair<상품, 상품개수>>와 같은 형태인데 Pair<상품, 상품개수> 를 cart로 표현하면 명칭이 헷갈릴 것 같습니다. 다른 네이밍을 고민해보면 어떨까요?

f-lab-martin commented on src/main/java/com/paytogether/cart/service/CartDto.java:
```diff
@@ -0,0 +1,26 @@
+
+@Getter
+@NoArgsConstructor
+public class CartDto {
```
ProductRequest와 동일한 멤버변수를 가지는데 CartDto로 표현하신 이유가 있으실까요?

tlqkrus012345 replied:
컨트롤러와 서비스에서 사용하는 dto를 구분하기 위해 CartDto 클래스를 구현했습니다!

f-lab-martin replied:
지금 작업하시는 부분에 두가지 도메인이 함께 있는 것으로 보입니다. cart, product 두가지 도메인이 있는 것으로 보입니다. 이러한 의존 관계를 어떻게 풀지 더 고민해보시면 좋을 것 같습니다.

f-lab-martin commented on src/main/java/com/paytogether/cart/service/CartDto.java:
```diff
@@ -0,0 +1,26 @@
+    return new CartDto(productId, quantity);
+  }
+
+  public int adjust(int quantity) {
```
사실상 setter와 동일한 역할인데 이 메소드를 쓰지 않는 방법은 없을까요?
그리고 DTO에 비즈니스로직이 있으면 어떤 단점이 있을까요?

tlqkrus012345 replied:
메소드를 사용하지 않고 새로운 CartDto를 생성하는 방법으로 수정했습니다!

DTO 클래스에 비즈니스 로직이 포함되면 멀티 스레드 환경에서 스레드 세이프하지 않다는 단점이 생깁니다.
DTO는 계층 간 데이터 전달을 목적으로 사용하는 객체이므로 특정 계층에서 해당 데이터의 불변성이 보장되지 않으면 코드의 신뢰성과 데이터의 정합성을 보장하지 못합니다.

f-lab-martin commented on src/main/java/com/paytogether/cart/service/CartService.java:
```diff
@@ -0,0 +1,95 @@
+  private final ObjectMapper objectMapper;
+
+  public CartResult addNewProduct(Long memberId, Long productId, int quantity) {
+    validateMemberId(memberId);
```
멤버가 존재하는지 보는 것보다 로그인 여부를 확인하는게 어떨까요? 
그리고 로그인 여부를 확인하는 로직이 CartService에 있어야할까요?

tlqkrus012345 replied:
로그인 여부를 확인하는 로직은 모든 로직의 공통된 관심사로 AOP를 적용하여 장바구니 기능과 로그인 여부 확인 기능을 분리하겠습니다! 해당 기능은 다음 pr에 요청드려도 될까요?

f-lab-martin replied:
넵 다음 PR에 추가해주세요~!

f-lab-martin replied:
추가로 장바구니에 상품을 담을 때, 실제로 존재하는 상품 ID인지 체크하는 로직이 추가로 필요해 보입니다. product 도메인과 관련된 서비스들도 필요해보이구요!

f-lab-martin commented on src/main/java/com/paytogether/cart/service/CartService.java:
```diff
@@ -0,0 +1,95 @@
+@Slf4j
+@RequiredArgsConstructor
+@Service
+public class CartService {
```
서비스 레이어에 비즈니스로직보다는 CartRepository에 가까운 로직들로 구성된 것으로 보입니다. 
레포지토리 레이어로 해당 로직을 분리하고 CartService에서는 비즈니스 로직들만 들어가도록 리팩토링해보면 어떨까요?

f-lab-martin commented on src/main/java/com/paytogether/member/service/HttpSessionManager.java:
```diff
@@ -8,22 +8,25 @@
 
-  public Long getMemberId() {
-    return (Long) httpSession.getAttribute(LOGIN_MEMBER);
+  public Long getMemberId(String key) {
+    return (Long) httpSession.getAttribute(key);
+  }
```
아래 로직과 중복으로 보입니다. 어떻게 수정할 수 있을까요?
```
public Object get(String key) {
    return httpSession.getAttribute(key);
  }

```

tlqkrus012345 replied:
중복을 제거하기 위해 매개변수에 Class<?> 를 추가하여 제너릭 타입으로 반환하게 수정했습니다!

f-lab-martin commented on src/test/java/com/paytogether/cart/service/CartServiceTest.java:
```diff
@@ -0,0 +1,82 @@
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+@SpringBootTest
+class CartServiceTest {
```
CartController 단위테스트도 같이 작성해보면 어떨까요?

tlqkrus012345 committed:
refactor: restful 설계 원칙

- 단수 -> 복수 (member -> members)
- clear 동사 제거
- memerId 등 메소드 파라미터 식별을 위해 members/{memberId}로 변경
- http 파일 경로 변수 명시화

Resolves: #8

tlqkrus012345 committed:
refactor: 모든 요청 경로 permitall 임시 처리

Resolves: #8

tlqkrus012345 committed:
refactor: dto -> record

- 데이터를 전송하는 객체를 표현하기 위해 record로 변경

Resolves: #8

tlqkrus012345 committed:
refactor: cart -> carts

Resolves: #8

tlqkrus012345 committed:
refactor: List<CartProductDto> cart -> cartProductDtoList 변수명 변경

Resolves: #8

tlqkrus012345 committed:
refacotr: dto 클래스 수정자 메소드

- dto 클래스에 비즈니스 로직을 제거하고 새로운 클래스를 할당하는 방법으로 변경

Resolves: #8

tlqkrus012345 committed:
refactor: CartService에서 저장하는 기능 CartRepository로 분리

- 서비스와 레포지토리 책임 분리

Resolves: #8

tlqkrus012345 committed:
test: CartService에 회원id 매개변수 제거

Resolves: #8

tlqkrus012345 committed:
refactor: 회원id 제거

Resolves: #8

tlqkrus012345 committed:
refactor: 메소드 이름 변경

Resolves: #8

tlqkrus012345 committed:
refactor: 메소드 이름 변경

Resolves: #8

tlqkrus012345 committed:
test: CartController

- 장바구니 상품 추가, 제거, 비우기, 조회 성공 테스트

Resolves: #8

tlqkrus012345 committed:
refactor: get(), getMemberId() 중복제거

- 세션에서 값을 조회하는 기능이 동일하고 반환 타입만 달라 제너릭 타입으로 반환하게 수정

Resolves: #8

tlqkrus012345 committed:
refactor: get() 수정

Resolves: #8

tlqkrus012345 committed:
test: getMemberId -> get() 수정

Resolves: #8

tlqkrus012345 committed:
test: member -> members url 변경

Resolves: #8

f-lab-martin commented on src/main/java/com/paytogether/cart/repository/CartRepository.java:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  public Map<String, CartProductDto> findCart() {
+    Map<String, CartProductDto> cart = httpSessionManager.get(REDIS_CART_KEY, Map.class);
+    return cart != null ? cart : new HashMap<>();
```
이전에는 HashOperations을 사용해서 구현하셨는데 바꾸신 이유가 있으실까요? 
두가지 방법의 장단점을 정리해보고, 현재 도메인의 요구사항을 고려했을 때, 어떤 방향을 선택하는 게 유리할지 정리해보면 어떨까요?

f-lab-martin reviewed (approved):
제가 추가로 남긴 코멘트는 다음 PR에 반영해주세요~!

tlqkrus012345 committed:
refactor: dto 클래스 record 적용

Resolves: #8

tlqkrus012345 committed:
test: record 적용에 따라 메소드 변경

Resolves: #8

tlqkrus012345 committed:
refactor: record 적용

Resolves: #8

tlqkrus012345 committed:
test: record 적용

Resolves: #8

tlqkrus012345 committed:
feat: product entity

Resolves: #8

tlqkrus012345 committed:
feat: product 예외

Resolves: #8

tlqkrus012345 committed:
test: 세션에서 회원 아이디 가져오는 방식 수정

Resolves: #8

tlqkrus012345 committed:
refactor: 세션에서 회원 아이디 가져오는 메소드

- 레디스에 저장할 떄 String 타입으로 저장
- get() -> getMemberId() 로 수정

Resolves: #8

tlqkrus012345 committed:
chore: data.sql 삽입 설정

Resolves: #8

tlqkrus012345 committed:
refactor: 장바구니 저장 방식

- httpsession을 이용하여 맵 형태로 저장을 redistemplate으로 변경

Resolves: #8

tlqkrus012345 committed:
test: dto 변경

Resolves: #8

tlqkrus012345 committed:
test: 사용하지 않는 메소드 제거

Resolves: #8

tlqkrus012345 committed:
feat: 생성자 추가

Resolves: #8

tlqkrus012345 committed:
feat: product 데이터

Resolves: #8

tlqkrus012345 committed:
refactor: productId

Resolves: #8
